### PR TITLE
Fix hygiene in objective macro

### PR DESCRIPTION
**Pull request by blegat**
Fixes the issue reported in https://discourse.julialang.org/t/undefvarerror-in-jump-constraint/15701/3

### CODE REVIEW

**Comment by mlubin** [comment]
Needs a test.

Can we make sure all macros work instead of fixing them one by one?

Does this affect 0.18.3?

**Comment by blegat** [comment]
> Can we make sure all macros work instead of fixing them one by one?

Yes, I am working on a better fix that defines two new functions that should be called at the end of every macros. The fix in getloopedcode wasn't sufficient

> Does this affect 0.18.3?

Yes, we will need to backport this fix

File: src/macros.jl [hunk 1]
```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
     end
 end
 
+"""
+    macro_return(model, code, variable)
+
+Return a code that runs `code` in a local scope which returns the value of `variable`.
```
Comment (mlubin) [inline src/macros.jl]: Nit: Return a block of code that 1. runs the code block `code` in a local scope and 2. returns the value of a local variable named `variable`. `variable` must reference a variable defined by `code`.

File: src/macros.jl [hunk 2]
```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
...
+    end
+end
+
+"""
+    macro_assign_and_return(model, code, variable, registerfun::Function,
+                            quotvarname, escvarname, final_variable=variable)
+
+Return runs `code` in a local scope which returns the value of `variable`
+and then assign `final_variable` to `escvarname`.
+If `registerfun` is given, it registers the value returned by the local scope
```
Comment (mlubin) [inline src/macros.jl]: `registerfun` is a function passed in. How do we know that it "registers the value returned by the local scope"?
The docstring should either provide the expected signature or call a concrete function.

File: src/macros.jl [hunk 3]
```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
...
+                                 registerfun::Union{Nothing, Function}=nothing,
+                                 model=nothing,
+                                 quotvarname=nothing)
+    macro_code = macro_return(code, variable)
+    return quote
+        $variable = $macro_code
+        $(if registerfun !== nothing
+              :($registerfun($model, $quotvarname, $variable))
+          end)
+        # escvarname should be set in the scope calling the macr
```
Comment (mlubin) [inline src/macros.jl]: Typo: "macro"

File: src/macros.jl [hunk 4]
```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
...
+        end
+    end
+end
+
+"""
+    macro_assign_and_return(model, code, variable, registerfun::Function,
+                            quotvarname, escvarname, final_variable=variable)
+
+Return runs `code` in a local scope which returns the value of `variable`
+and then assign `final_variable` to `escvarname`.
```
Comment (mlubin) [inline src/macros.jl]: Could we take the `varname` and `quotvarname` as a single symbol and escape/quote it as needed?

Comment (blegat) [reply src/macros.jl]: Indeed, I was able to get rid of `quotvarname` and `escvarname`. It makes things more readable to isolate this technical details in macro_assign_and_return :)

File: src/macros.jl [hunk 5]
```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
...
+            $code
+            $variable
+        end
+    end
+end
+
+"""
+    macro_assign_and_return(code, variable, name;
+                            final_variable=variable,
+                            registerfun::Union{Nothing, Function}=nothing,
```
Comment (mlubin) [inline src/macros.jl]: Style nit: `register_fun`

File: test/Untitled.ipynb [hunk 6]
```diff
@@ -0,0 +1,222 @@
+{
```
Comment (mlubin) [inline test/Untitled.ipynb]: Hmm?

**Comment by carlocab** [comment]
Hi there, I'm the one who reported the issue at [the discourse page]( https://discourse.julialang.org/t/undefvarerror-in-jump-constraint/15701/3). I can see that this PR has been merged into master, but I'm getting the exact same problems as reported on discourse when I use the master branch.

**Pull request merged**